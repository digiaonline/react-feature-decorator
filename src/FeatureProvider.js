// @flow

import {Component, Children} from 'react'
import featuresShape from './featuresShape'
import type {FeatureName} from './types'

type Props = {
  features: Array<FeatureName>,
  enabledFeatures: Array<FeatureName>,
  children: any,
}

class FeatureProvider extends Component<Props, void> {
  static defaultProps = {
    features: [],
    enabledFeatures: [],
  }

  static childContextTypes = {
    features: featuresShape.isRequired,
  }

  getChildContext() {
    return {
      features: {
        isDefined: this.isDefined,
        isEnabled: this.isEnabled,
        tryIsDefined: this.tryIsDefined,
      },
    }
  }

  isDefined = (featureName: FeatureName) => {
    return this.props.features.indexOf(featureName) !== -1
  }

  tryIsDefined = (featureName: FeatureName) => {
    if (!this.isDefined(featureName)) {
      throw `Feature '${featureName}' is not defined!`
    }
  }

  isEnabled = (featureName: FeatureName) => {
    this.tryIsDefined(featureName)

    return this.props.enabledFeatures.indexOf(featureName) !== -1
  }

  render() {
    return this.props.children ? Children.only(this.props.children) : null
  }
}

export default FeatureProvider
