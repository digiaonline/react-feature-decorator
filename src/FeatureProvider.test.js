import React from 'react'
import renderer from 'react-test-renderer'
import FeatureProvider from './FeatureProvider'

const createProvider = props => renderer.create(<FeatureProvider {...props} />)

test('Renders only children', () => {
  const tree = createProvider({
    children: <div/>
  })
 
   expect(tree.toJSON()).toMatchSnapshot()
 })

test('Renders without children', () => {
 const tree = createProvider()

  expect(tree.toJSON()).tobeNull
})

test('Recognizes defined features', () => {
  const tree = createProvider({features: ['foo']})
  const component = tree.getInstance();
 
  expect(component.isDefined('foo')).toBeTruthy
  expect(component.isDefined('bar')).toBeTruthy
})

test('Recognizes enabled features', () => {
  const tree = createProvider({features: ['foo', 'bar'], enabledFeatures: ['foo']})
  const component = tree.getInstance();
 
  expect(component.isEnabled('foo')).toBeTruthy
  expect(component.isEnabled('bar')).toBeFalsy
  expect(() => component.isEnabled('baz')).toThrow(`Feature 'baz' is not defined!`)
})