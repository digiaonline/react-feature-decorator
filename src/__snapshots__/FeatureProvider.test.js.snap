// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders only children 1`] = `<div />`;
